drv/lpc55-sprot-server/src/main.rs

```diff
@@ -0,0 +1,420 @@
+        &server.io.rx_buf[0..0],
+        &mut server.io.tx_buf[..],
+        server.status,
+    );
+
+    loop {
+        server.io.pio(transmit);
+        let iostat = if server.io.rxcount == 0
+            && !server.io.underrun
+            && !server.io.overrun
+        {
+            IoStatus::Flush
+        } else {
+            IoStatus::IOResult {
+                underrun: server.io.underrun,
+                overrun: server.io.overrun,
+            }
+        };
+        transmit = match server.handler.handle(
+            transmit, // true if previous loop transmitted.
+            iostat,
+            &server.io.rx_buf[0..server.io.rxcount],
+            &mut server.io.tx_buf[..],
+            server.status,
+        ) {
+            Some(_txlen) => {
+                ringbuf_entry!(Trace::HandlerReturnSize(_txlen));
+                true
+            }
+            None => {
+                ringbuf_entry!(Trace::HandleReturnNone);
+                false
+            }
+        }
+    }
+}
+
+impl IO {
+    /// Wait for chip select to be asserted then service the FIFOs until end of frame.
+    /// Returns false on spurious interrupt.
+    fn pio(&mut self, transmit: bool) {
+        ringbuf_entry!(Trace::Pio(transmit));
+        let tx_end = self.tx_buf.len(); // Available bytes and trailing zeros
+        let rx_end = self.rx_buf.len(); // All of the available bytes
+        self.txcount = 0;
+        self.rxcount = 0;
+        self.overrun = false;
+        self.underrun = false;
+
+        if !transmit {
+            // Ensure that unused Tx buffer is zero-filled.
+            self.tx_buf.fill(0);
+        }
+
+        // Prime FIFOWR in order to be ready for start of frame.
+        //
+        // All our interrupts are left enabled for the sake of simplicity.
+        // The downside is that the following drain will elicit a spurious
+        // interrupt. But, that interrupt will occur before the start of any
+        // frame when we are not in a realtime situation.
+        self.spi.drain_tx(); // FIFOWR is now empty; we'll get an interrupt.
+        loop {
+            if self.txcount >= tx_end || !self.spi.can_tx() {
+                break;
+            }
+            let b = self.tx_buf[self.txcount];
+            self.spi.send_u8(b);
+            self.txcount += 1;
+        }
+
+        if transmit {
+            ringbuf_entry!(Trace::Transmit(self.txcount, tx_end));
+            self.gpio.set_val(ROT_IRQ, Value::Zero).unwrap_lite();
+            ringbuf_entry!(Trace::RotIrqAssert);
+        }
+
+        // TODO: SP RESET needs to be monitored, otherwise, any
+        // forced looping here could be a denial of service attack against
+        // observation of SP resetting. SP resetting without invalidating
+        // security related state means a compromised SP could operate using
+        // the trust gained in the previous session.
+        // Upper layers may mitigate that, but check on it.
```

Comment:
Do we expect SP RESET to be monitored here in this driver or another driver?

Reply:
This is one place it could be monitored, but only because it is monitoring other things that come from the SP.
My thinking is that sprockets is linked into this task and may have state that needs to be invalidated.
There will be work to do elsewhere, so I don't think that sprockets is forcing the decision.